linkerd/app/core/src/errors/respond.rs

```diff
@@ -76,76 +121,138 @@ impl<ReqB, RspB: Default + hyper::body::HttpBody>
-                    debug!(?code, "Handling error with gRPC status");
-                    return Ok(rsp);
-                }
+        let span = info_span!(
+            "rescue",
+            client.addr = %self.client
+                .as_ref()
+                .map(|ClientHandle { addr, .. }| *addr)
+                .unwrap_or_else(|| {
+                    tracing::debug!("Missing client address");
+                    ([0, 0, 0, 0], 0).into()
+                })
+        );
 
-                let rsp = set_http_status(builder, &*error)
-                    .version(self.version)
-                    .header(http::header::CONTENT_LENGTH, "0")
-                    .body(ResponseBody::default())
-                    .expect("error response must be valid");
-                let status = rsp.status();
-                debug!(%status, version = ?self.version, "Handling error with HTTP response");
-                Ok(rsp)
+        let SyntheticHttpResponse {
+            grpc_status,
+            http_status,
+            close_connection,
+            message,
+        } = span.in_scope(|| {
+            tracing::info!(%error, "Request failed");
+            self.rescue.rescue(error)
+        })?;
+
+        if close_connection {
+            if let Some(ClientHandle { close, .. }) = self.client.as_ref() {
+                close.close();
+            } else {
+                tracing::debug!("Missing client handle");
             }
         }
+
+        let rsp = if self.is_grpc {
+            Self::grpc_response(grpc_status, &*message)
+        } else {
+            Self::http_response(self.version, http_status, &*message, close_connection)
+        };
+
+        Ok(rsp)
+    }
+}
+
+impl<R> Respond<R> {
+    fn http_response<B: Default>(
+        version: http::Version,
+        status: http::StatusCode,
+        message: &str,
+        close_connection: bool,
+    ) -> http::Response<B> {
+        info!(%status, ?version, %message, close_connection, "Handling error on HTTP connection");
+        let mut rsp = http::Response::builder()
+            .status(status)
+            .version(version)
+            .header(http::header::CONTENT_LENGTH, "0")
+            .header(
+                L5D_PROXY_ERROR,
+                HeaderValue::from_str(message).unwrap_or_else(|error| {
+                    warn!(%error, "Failed to encode error header");
+                    HeaderValue::from_static("Unexpected error")
+                }),
+            );
+
+        if close_connection && version == http::Version::HTTP_11 {
+            rsp = rsp.header(http::header::CONNECTION, "close");
+        }
+
+        rsp.body(B::default())
+            .expect("error response must be valid")
+    }
+
+    fn grpc_response<B: Default>(code: tonic::Code, message: &str) -> http::Response<B> {
+        info!(%code, %message, "Handling error on gRPC connection");
+
+        http::Response::builder()
```

Comment:
nit/TIOLI (possibly a premature optimization): we always add exactly 5 headers to the headermap here, so I wonder if we want to consider pre-allocating enough capacity for them? that way, we know it won't allocate multiple times...
```suggestion
        let mut builder = http::Response::builder();
        // We always add 5 headers, so preallocate enough capacity for them.
        builder
            .headers_mut()
            .expect("builder will not have an error")
            .reserve(5);
        builder
```